init/eessi_archdetect.sh

```diff
@@ -0,0 +1,81 @@
+#!/bin/bash
+
+# current pathways implemented in EESSI
+# x86_64/generic
+# x86_64/intel/haswell
+# x86_64/intel/skylake_avx512
+# x86_64/amd/zen2
+# x86_64/amd/zen3
+# aarch64/generic
+# ppc64le/generic
+# ppc64le/power9le
+
+ARGUMENT=${1:-none}
+
+cpupath () {
+  # let the kernel tell base machine type
+  MACHINE_TYPE=${EESSI_MACHINE_TYPE:-$(uname -m)}
+
+  # fallback path
+  CPU_PATH="${MACHINE_TYPE}/generic"
+
+  if [ ${MACHINE_TYPE} == "aarch64" ]; then
+    echo ${CPU_PATH}
+    exit
+  fi
+
+  if [ ${MACHINE_TYPE} == "ppc64le" ]; then
+    PROC_CPUINFO=${EESSI_PROC_CPUINFO:-/proc/cpuinfo}
+    CPU_FLAGS=$(grep -m 1 -i ^cpu ${PROC_CPUINFO})
+    [[ $CPU_FLAGS =~ .*POWER9* ]] && HAS_POWER9=true
+
+    [[ ${HAS_POWER9} ]] && CPU_PATH=${MACHINE_TYPE}/power9le
+    echo ${CPU_PATH}
+    exit
+  fi
+
+  if [ ${MACHINE_TYPE} == "x86_64" ]; then
+    # check for vendor info, if available, for x86_64
+    PROC_CPUINFO=${EESSI_PROC_CPUINFO:-/proc/cpuinfo}
+    CPUINFO_VENDOR_FLAG=$(grep -m 1 -i ^vendor ${PROC_CPUINFO})
+    [[ $CPUINFO_VENDOR_FLAG =~ .*GenuineIntel* ]] && CPU_VENDOR=intel
+    [[ $CPUINFO_VENDOR_FLAG =~ .*AuthenticAMD* ]] && CPU_VENDOR=amd
+
+    CPU_FLAGS=$(grep -m 1 -i ^flags ${PROC_CPUINFO})
+    [[ $CPU_FLAGS =~ .*avx2* ]] && HAS_AVX2=true
+    [[ $CPU_FLAGS =~ .*fma* ]] && HAS_FMA=true
+    [[ $CPU_FLAGS =~ .*avx512f* ]] && HAS_AVX512F=true
+    [[ $CPU_FLAGS =~ .*avx512vl* ]] && HAS_AVX512VL=true
+    [[ $CPU_FLAGS =~ .*avx512ifma* ]] && HAS_AVX512IFMA=true
+    [[ $CPU_FLAGS =~ .*avx512_vbmi2* ]] && HAS_AVX512_VBMI2=true
+    [[ $CPU_FLAGS =~ .*avx512_vnni* ]] && HAS_AVX512_VNNI=true
+    [[ $CPU_FLAGS =~ .*avx512fp16* ]] && HAS_AVX512FP16=true
+    [[ $CPU_FLAGS =~ .*vaes* ]] && HAS_VAES=true
+
+    if [ ${CPU_VENDOR} == "intel" ]; then
+      [[ ${HAS_AVX2} ]] && [[ ${HAS_FMA} ]] && CPU_TYPE=haswell 
+      [[ ${HAS_AVX512F} ]] && CPU_TYPE=skylake_avx512
+      # [[ ${HAS_AVX512IFMA} ]] && [[ ${HAS_AVX512_VBMI2} ]] && CPU_TYPE=icelake_avx512
+      # [[ ${HAS_AVX512_VNNI} ]] && [[ ${HAS_AVX512VL} ]] && [[ ${HAS_AVX512FP16} ]] && CPU_TYPE=sapphire_rapids_avx512
+    elif [ ${CPU_VENDOR} == "amd" ]; then
+      [[ ${HAS_AVX2} ]] && [[ ${HAS_FMA} ]] && CPU_TYPE=zen2
```

Comment:
Maybe we should add a comment here to make this a bit easier to find when people look for `rome` or `milan`:

```shell
[[ ${HAS_AVX2} ]] && [[ ${HAS_FMA} ]] && CPU_TYPE=zen2  # Rome
[[ ${HAS_AVX2} ]] && [[ ${HAS_FMA} ]] && CPU_TYPE=zen3  # Milan, Milan-X